### PR TITLE
Query scorer refactoring

### PR DESCRIPTION
**Pull request by generall**
Current version of raw_scorer assumes, that there is only one way to evaluate score during search: with metric function.

This PR is aimed to extend the scope of `Query Score` - measure of similarity between vector and query, where query might be more than just a single vector.

### CODE REVIEW

File: lib/segment/src/vector_storage/raw_scorer.rs [hunk 1]
```diff
@@ -1,15 +1,27 @@
...
+/// ┌────────────────┐   ┌──────────────┐   ┌───────────────────┐
+/// │                │   │              │   │  - Cosine         │
+/// │       ┌─────┐  │   │     ┌─────┐  │   │  - Dot            │
+/// │       │     │◄─┼───┤     │     │◄─┼───┤  - Euclidean      │
+/// │       └─────┘  │   │     └─────┘  │   │                   │
+/// │                │   │              │   │                   │
+/// └────────────────┘   └──────────────┘   └───────────────────┘
+/// - Deletions          - Scoring logic    - Vector Distance
+/// - Stopping           - Query holding
+/// - Access patterns    - Complex Queries
```
Comment (generall) [inline lib/segment/src/vector_storage/raw_scorer.rs:23]: Main idea here

File: lib/api/src/grpc/google.protobuf.rs [hunk 2]
```diff
@@ -16,8 +16,8 @@ pub struct Struct {
 }
 /// `Value` represents a dynamically typed value which can be either
 /// null, a number, a string, a boolean, a recursive struct value, or a
-/// list of values. A producer of value is expected to set one of these
-/// variants. Absence of any variant indicates an error.
+/// list of values. A producer of value is expected to set one of that
+/// variants, absence of any variant indicates an error.
```
Comment (timvisee) [inline lib/api/src/grpc/google.protobuf.rs:20]: `these` was correct :smiley: 

```suggestion
/// list of values. A producer of value is expected to set one of these
/// variants, absence of any variant indicates an error.
```

File: lib/segment/src/vector_storage/raw_scorer.rs [hunk 3]
```diff
@@ -181,19 +195,17 @@ where
...
-                score: TMetric::similarity(&self.query, other_vector),
+                score: self.query_scorer.score_stored(point_id),
             });
         }
         scores
     }
 
     fn check_vector(&self, point: PointOffsetType) -> bool {
-        point < self.points_count
-            // Deleted points propagate to vectors; check vector deletion for possible early return
```
Comment (timvisee) [inline lib/segment/src/vector_storage/raw_scorer.rs:195]: ~Do we remove this bound check on purpose? IIRC we had trouble with this before, and using this bound check would give us more reliable results.~

We already do this bound check with `points_deleted`, nice!

**Review by timvisee — APPROVED** [review]
Nice work on decomposing this!

File: lib/segment/src/vector_storage/quantized/quantized_query_scorer.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+impl<TEncodedQuery, TEncodedVectors> QueryScorer
+    for QuantizedQueryScorer<'_, TEncodedQuery, TEncodedVectors>
+where
+    TEncodedVectors: quantization::EncodedVectors<TEncodedQuery>,
+{
+    fn score_stored(&self, idx: PointOffsetType) -> ScoreType {
+        self.quantized_data.score_point(&self.query, idx)
+    }
+
+    fn score(&self, v2: &[VectorElementType]) -> ScoreType {
```
Comment (IvanPleshkov) [inline lib/segment/src/vector_storage/quantized/quantized_query_scorer.rs:43]: This method is not used everywhere. I propose to remove it, some code can be removed too after this function

Comment (generall) [reply lib/segment/src/vector_storage/quantized/quantized_query_scorer.rs:43]: it is used in Async Scorer

**Review by IvanPleshkov — APPROVED** [review]

**Pull request merged**